drv/sidecar-front-io-api/src/controller.rs

```diff
@@ -0,0 +1,56 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::Addr;
+use drv_fpga_api::*;
+
+static COMPRESSED_BITSTREAM: &[u8] =
+    include_bytes!(concat!(env!("OUT_DIR"), "/ecp5.bin.rle"));
+
+pub struct FrontIOController {
+    fpga: Fpga,
+    user_design: FpgaUserDesign,
+}
+
+impl FrontIOController {
```

Comment:
Is it worth combining `FrontIOController` and `MainboardController`, which share basically the same logic?

Reply:
I want to keep them separate structs, because they convey a different context. But I agree that duplicate code is meh, so I've started moving most into `fpga-api` where it can be used by anything.